### PR TITLE
Fix upgrade tests

### PR DESCRIPTION
**Pull request by QGarchery**

### CODE REVIEW

File: test/prod/ProdTest.sol [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    // Needed because AAVE packs the balance struct.
+    function dealAave(address who, uint104 amount) public {
+        // The slot of the balance struct "_balances" is 0.
+        bytes32 slot = keccak256(abi.encode(who, uint256(0)));
+        bytes32 initialValue = vm.load(aave, slot);
+        // The balance is stored in the first 104 bits.
+        bytes32 finalValue = ((initialValue >> 104) << 104) | bytes32(uint256(amount));
+        vm.store(aave, slot, finalValue);
+        require(ERC20(aave).balanceOf(who) == uint256(amount));
+    }
```
Comment (Rubilmax) [inline test/prod/ProdTest.sol:24]: This was also handled in aaveV3 optimizer this way: https://github.com/morpho-org/morpho-aavev3-optimizer/blob/5acc62c04a5e8032d4cce399dea31442ef6c2ad2/test/helpers/ForkTest.sol#L188-L198

At some point (I don't remember where), we even overridden forge-std's `deal` function to take into account the AAVE edge case (i really like this design)

I don't think creating a `ProdTest` layer for this is interesting: it adds a layer whereas we know that all tests in this folder are production tests (they all inherit from `prod/TestSetup`)

Comment (QGarchery) [reply test/prod/ProdTest.sol:24]: > This was also handled in aaveV3 optimizer this way: https://github.com/morpho-org/morpho-aavev3-optimizer/blob/5acc62c04a5e8032d4cce399dea31442ef6c2ad2/test/helpers/ForkTest.sol#L188-L198

Interesting, I did not know that. Do you think we should change the code to match the one you linked ?

> I don't think creating a ProdTest layer for this is interesting: it adds a layer whereas we know that all tests in this folder are production tests (they all inherit from prod/TestSetup)

This is what I had initially in the first few commits of this PR, see [this commit](https://github.com/morpho-org/morpho-optimizers/pull/1669/commits/738ad6d71d05cd95c6330c8f13890ed66b979233), but then I realized that there are 2 different `TestSetup`: one for `aave-v2` and one for `compound`. So it would mean duplicating the function, do you think we should do it ?

Comment (Rubilmax) [reply test/prod/ProdTest.sol:24]: > Do you think we should change the code to match the one you linked ?

If your code works, I don't have a strong argument in favor of mine ^^ so keeping yours is fine

> So it would mean duplicating the function, do you think we should do it ?

Ah right, missed this! `ProdTest` is inherited by `TestSetup`. Well then i would only argue towards overriding forge-std's `deal` as in the code i referenced, keeping your workaround or mine (i don't care) ; so we don't have to check if the token is AAVE outside of this function

It can be useful when fuzzing the token but perhaps we don't do it in this repo. In which case let's just keep your version because this code is getting archived anyway

Comment (QGarchery) [reply test/prod/ProdTest.sol]: Ok I changed both because they seem cleaner. One little caveat is that the AAVE token is not supposed to work with values greater than `type(uint104).max`, so we would have to distinguish the cases in fuzzing still (with the current solution at least)

File: test/prod/compound/TestLifecycle.t.sol [hunk 2]
```diff
@@ -143,10 +141,9 @@ contract TestLifecycle is TestSetup {
             supply.morphoUnderlyingBalanceBefore,
             string.concat(supply.market.symbol, " morpho balance")
         );
-        assertApproxEqAbs(
-            ICToken(supply.market.poolToken).balanceOfUnderlying(address(morpho)),
-            supply.morphoPoolSupplyBefore + supply.position.pool,
-            10,
+        assertTrue(
+            ICToken(supply.market.poolToken).balanceOfUnderlying(address(morpho)) >=
+                supply.morphoPoolSupplyBefore + supply.position.pool,
```
Comment (Rubilmax) [inline test/prod/compound/TestLifecycle.t.sol:146]:
```suggestion
        assertGe(
            ICToken(supply.market.poolToken).balanceOfUnderlying(address(morpho)),
                supply.morphoPoolSupplyBefore + supply.position.pool,
```

to also log values in case of error

**Review by Rubilmax — DISMISSED** [review]

**Comment by Rubilmax** [comment]
Upgrade test fails on aave v2

**Review by Rubilmax — APPROVED** [review]

**Review by Jean-Grimal — APPROVED** [review]

**Comment by MerlinEgalite** [comment]
Tests are still failing cc @QGarchery

**Review by Rubilmax — APPROVED** [review]

File: test/prod/ProdTest.sol [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import {WadRayMath} from "@morpho-dao/morpho-utils/math/WadRayMath.sol";
+
+import {BaseConfig} from "config/BaseConfig.sol";
+import "@forge-std/console.sol";
+import "@forge-std/Test.sol";
+
+contract ProdTest is Test, BaseConfig {
+    function setUp() public virtual {
+        console.log("Testing at block", block.number);
+    }
```
Comment (QGarchery) [inline test/prod/ProdTest.sol:17]: Do you prefer having a dedicated dummy test that logs the block number, in order to reduce the noise of having a log in each test ?

Comment (MathisGD) [reply test/prod/ProdTest.sol:17]: yes good idea

File: test/prod/compound/TestUpgradeLens.t.sol [hunk 4]
```diff
@@ -124,7 +124,7 @@ contract TestUpgradeLens is TestSetup {
                 assertApproxEqAbs(
                     lens.getCurrentUserSupplyRatePerBlock(supplyMarket.poolToken, address(user)),
                     expectedSupplyRate,
-                    1,
+                    1e16,
```
Comment (MerlinEgalite) [inline test/prod/compound/TestUpgradeLens.t.sol:127]: this very large no?

Comment (QGarchery) [reply test/prod/compound/TestUpgradeLens.t.sol:127]: Yes, but it's the same as the test in aave-v2. I'll look into it

Comment (QGarchery) [reply test/prod/compound/TestUpgradeLens.t.sol]: I investigated and found that:
- the number of matching iterations is correctly set to 1
- the difference comes from the pool rates (virtual vs actual) and is of similar size
- the difference is way lower to what we had before (so I decreased the deltas to 1e8) because of the fix to the borrow amount being able to include reserves

For these reasons, I'm marking this comment as resolved

**Review by MerlinEgalite — APPROVED** [review]

File: test/prod/compound/setup/TestSetup.sol [hunk 5]
```diff
@@ -200,9 +200,7 @@ contract TestSetup is Config, Test {
                     Math.min(
                         Math.min(
                             (_market.maxBorrows - _market.totalBorrows) / 2,
-                            _market.underlying == wEth
-                                ? cEth.balance
-                                : ERC20(_market.underlying).balanceOf(_market.poolToken)
+                            poolToken.getCash() - poolToken.totalReserves()
```
Comment (Rubilmax) [inline test/prod/compound/setup/TestSetup.sol:203]: Nice catch

**Review by Rubilmax — APPROVED** [review]

File: test/prod/ProdTest.sol [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import "@forge-std/Test.sol";
+
+contract ProdTest is Test, BaseConfig {
+    // Show block number for reproducibility.
+    function testShowBlockNumber() public view {
+        console.log("Testing at block", block.number);
+    }
+
+    // Needed because AAVE packs the balance struct.
+    function deal(
```
Comment (MathisGD) [inline test/prod/ProdTest.sol:21]: why was it needed if we can just give type(uint104).max ?

```solidity
         // Only 104 bits are used for the balance in the AAVE balance slot
         deal(aave, address(this), type(uint104).max);
```

Comment (QGarchery) [reply test/prod/ProdTest.sol:21]: We can't do `type(uint104).max` only because it would revert. This is a recent forge upgrade that prevents doing `deal` when the target slot is a struct
